builtin/plugins/dposv3/dpos.go

```diff
@@ -1323,6 +1332,51 @@ func ShiftDowntimeWindow(ctx contract.Context, currentHeight int64, candidates [
 	return nil
 }
 
+func jailOfflineValidators(ctx contract.Context, cachedDelegations *CachedDposStorage) error {
+	validators, err := ValidatorList(ctx)
+	if err != nil {
+		return logDposError(ctx, err, "jailOfflineValidators could not load validator list")
+	}
+
+	state, err := loadState(ctx)
+	if err != nil {
+		return logDposError(ctx, err, "jailOfflineValidators could not load state")
+	}
+
+	params := state.Params
+
+	for _, validator := range validators {
+		candidate := GetCandidateByPubKey(ctx, validator.PubKey)
+
+		if candidate == nil {
+			ctx.Logger().Info("Attempted to check downtime of a validator no longer on candidates list.", "validator", validator)
+			continue
+		}
+
+		candidateAddress := loom.UnmarshalAddressPB(candidate.Address)
+		statistic, _ := GetStatistic(ctx, candidateAddress)
+		if statistic != nil {
+			// if a validator misses all blocks in the last 4 periods, jail them
+			if statistic.RecentlyMissedBlocks&0xFFFF == params.DowntimePeriod &&
```

Comment:
use the `getDowntimeRecord` function here.

Reply:
This logic is deprecated. Please take a look at the new one. Yes I can use `getDowntimeRecord` but that is going to another hit to the db, so I think it is better to use something we already have in the memory. Maybe we can create a utility function for handling bit manipulation so that we don't have to copy & paste the logic.

Reply:
We definitely don't want to reproduce this code so factor out the bit-munging from `getDowntimeRecord` into a new function that accepts a statistic, like george suggested below, and then have both the code here and `getDowntimeRecord call this new function.